cleanlab/outlier.py

```diff
@@ -0,0 +1,534 @@
+                    f"OOD Object needs to be fit on features first. Call fit() or fit_scores() before this function."
+                )
+            else:
+                scores, _ = _get_ood_predictions_scores(
+                    pred_probs, **self._get_params(self.OOD_PARAMS)
+                )
+
+        return scores
+
+    def _get_params(self, param_keys) -> dict:
+        """
+        Helper method to get function specific dictionary of parameters (i.e. only those in param_keys).
+        """
+        return {k: v for k, v in self.params.items() if k in param_keys}
+
+    @staticmethod
+    def _assert_valid_params(params, param_keys):
+        """
+        Helper method to check passed in params valid and get list of parameters in param that are not in param_keys.
+        """
+        if params is not None:
+            wrong_params = list(set(params.keys()).difference(set(param_keys)))
+            if len(wrong_params) > 0:
+                raise ValueError(
+                    f"Passed in params dict can only contain {param_keys}. Remove {wrong_params} from params dict."
+                )
+
+    @staticmethod
+    def _assert_valid_inputs(features, pred_probs):
+        """
+        Helper method to check features and pred_prob inputs are valid. Throws error if not.
+        """
+        if features is None and pred_probs is None:
+            raise ValueError(
+                f"Not enough information to compute scores. Pass in either features or pred_probs."
+            )
+
+        if features is not None and pred_probs is not None:
+            raise ValueError(
+                f"Cannot fit to object to both features and pred_probs. Pass in either one or the other."
+            )
+
+        if features is not None and len(features.shape) != 2:
+            raise ValueError(
+                f"Feature array needs to be of shape (N, M), where N is the number of examples and M is the "
+                f"number of features used to represent each example. "
+            )
+
+    def _shared_fit(
+        self,
+        *,
+        features: Optional[np.ndarray] = None,
+        pred_probs: Optional[np.ndarray] = None,
+        labels: Optional[np.ndarray] = None,
+        verbose: bool = True,
+    ) -> Optional[np.ndarray]:
+        """
+        Shared fit functionality between `fit()` and `fit_score()`.
+        For details, refer to :py:func:`fit <cleanlab.outlier.OutOfDistribution.fit>`
+        or :py:func:`fit_score <cleanlab.outlier.OutOfDistribution.fit_score>`.
+        """
+        # Assert valid inputs and params
+        self._assert_valid_inputs(features, pred_probs)
+        scores = None  # If none scores are returned, fit was skipped
+
+        if features is not None:
+            if self.params["knn"] is not None:
+                # No fitting twice if knn object already fit
+                warnings.warn(
+                    f"Skipping fit. Object already fit on knn estimator.",
```

Comment:
"Object already fit on knn estimator"
->
"A KNN estimator has previously already been fit, call `score()` to use it, or create a new `OutOfDistribution` object to fit a different estimator."